### ISSUE TITLE
WASI: path_open regression in Rust 1.51 when using LTO

### ISSUE THREAD

**Issue by PiotrSikora** [rust-lang/rust#82758]
### Code

I tried this code:

```rust
fn main() {
    let _ = std::fs::File::open("/proxy-wasm-sandbox/regression.txt");
}
```
```shell
$ cargo +beta build --target wasm32-wasi --release
```

I expected to see this happen:

This is executed inside Envoy Proxy with preopen dir `/proxy-wasm-sandbox`, but hopefully the output is readable:
```
[host->vm] _start()
[vm->host] wasi_snapshot_preview1.fd_prestat_get(3, 1048568)
[vm<-host] wasi_snapshot_preview1.fd_prestat_get return: 0
[vm->host] wasi_snapshot_preview1.fd_prestat_dir_name(3, 1114128, 19)
[vm<-host] wasi_snapshot_preview1.fd_prestat_dir_name return: 0
[vm->host] wasi_snapshot_preview1.fd_prestat_get(4, 1048568)
[vm<-host] wasi_snapshot_preview1.fd_prestat_get return: 8
[vm->host] wasi_snapshot_preview1.path_open(3, 1, 1114320, 14, 0, 262667966, 262667966, 0, 1048520)
[vm<-host] wasi_snapshot_preview1.path_open return: 0
[vm->host] wasi_snapshot_preview1.fd_close(4)
[vm<-host] wasi_snapshot_preview1.fd_close return: 0
[host<-vm] _start return: void
```
The memory address and size passed in the call (`path_open(.., 1114320, 14, ..)`) correctly point to a `regression.txt`. 

Instead, this happened:
```
[host->vm] _start()
[vm->host] wasi_snapshot_preview1.fd_prestat_get(3, 1048568)
[vm<-host] wasi_snapshot_preview1.fd_prestat_get return: 0
[vm->host] wasi_snapshot_preview1.fd_prestat_dir_name(3, 1058576, 19)
[vm<-host] wasi_snapshot_preview1.fd_prestat_dir_name return: 0
[vm->host] wasi_snapshot_preview1.fd_prestat_get(4, 1048568)
[vm<-host] wasi_snapshot_preview1.fd_prestat_get return: 8
[vm->host] wasi_snapshot_preview1.path_open(3, 1, 1, 0, 0, 262667966, 262667966, 0, 1048520)
[vm<-host] wasi_snapshot_preview1.path_open return: 44
[host<-vm] _start return: void
```
The memory address and size passed in the call (`path_open(.., 1, 0, ..)`) are wrong.

### Version it worked on

It most recently worked on: Rust 1.50

`rustc --version --verbose`:
```
rustc 1.50.0 (cb75ad5db 2021-02-10)
binary: rustc
commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
commit-date: 2021-02-10
host: x86_64-unknown-linux-gnu
release: 1.50.0
```

### Version with regression

`rustc --version --verbose`:
```
rustc 1.51.0-beta.3 (b631c914c 2021-02-24)
binary: rustc
commit-hash: b631c914cdb5b74dac8f182ca96fdfd3e03ab319
commit-date: 2021-02-24
host: x86_64-unknown-linux-gnu
release: 1.51.0-beta.3
LLVM version: 11.0.1
```

@rustbot modify labels: +regression-from-stable-to-beta -regression-untriaged

**Comment by PiotrSikora** [comment]
Those PRs look like a possible candidates: https://github.com/rust-lang/rust/pull/79578, https://github.com/WebAssembly/wasi-libc/pull/214.

Perhaps there is a bug in my code, but it works with Rust 1.50, and the preopens are populated correctly even in Rust 1.51 (otherwise it wouldn't reach `path_open`), so there aren't too many places left to make a mistake.

**Comment by PiotrSikora** [comment]
I think I found the culprit.

This happens only in the release builds with LTO enabled, where `__wasilibc_find_relpath_alloc` gets optimized out:
```
  (func $undefined:__wasilibc_find_relpath_alloc (type $9)
    (param $0 i32)
    (param $1 i32)
    (param $2 i32)
    (param $3 i32)
    (param $4 i32)
    (result i32)
    unreachable
    )
```

**Comment by PiotrSikora** [comment]
Confirmed: if I "use" `__wasilibc_find_relpath_alloc`, so that it doesn't get optimized out, then everything works fine:

```rust
use libc;

extern "C" {
    fn __wasilibc_find_relpath_alloc(
        path: *const libc::c_char,
        abs_prefix: *mut *const libc::c_char,
        relative_path: *mut *const libc::c_char,
        relative_path_len: libc::size_t,
        can_realloc: libc::c_int,
    ) -> libc::c_int;
}

fn main() {
    unsafe {
        let _ = __wasilibc_find_relpath_alloc(
            0 as *const libc::c_char,
            0 as *mut *const libc::c_char,
            0 as *mut *const libc::c_char,
            0,
            0,
        );
    }
    let _ = std::fs::File::open("/proxy-wasm-sandbox/regression.txt");
}
```

**Comment by alexcrichton** [comment]
@PiotrSikora can you clarify how you're building your project? Are you using an external libc or the one shipped with libstd?

**Comment by PiotrSikora** [comment]
```
$ find . -type f
```
```
./src/main.rs
./Cargo.toml
```
```
$ cat Cargo.toml
```
```
[package]
name = "regress"
version = "0.1.0"
authors = ["Piotr Sikora <piotrsikora@google.com>"]
edition = "2018"

[profile.release]
lto = true
opt-level = 3
panic = "abort"
```
```
$ cat src/main.rs
```
```
fn main() {
    let _ = std::fs::File::open("/proxy-wasm-sandbox/regression.txt");
}
```
```
$ cargo +beta --version && cargo +beta build --target wasm32-wasi --release
```
```
cargo 1.51.0-beta (929458982 2021-02-22)
   Compiling regress v0.1.0 (/tmp/regress)
    Finished release [optimized] target(s) in 1.85s
```
```
$ wasm-dis target/wasm32-wasi/release/regress.wasm | grep -A2 __wasilibc_find_relpath_alloc
```
```
 (func $undefined:__wasilibc_find_relpath_alloc (param $0 i32) (param $1 i32) (param $2 i32) (param $3 i32) (param $4 i32) (result i32)
  (unreachable)
 )
--
     (call $undefined:__wasilibc_find_relpath_alloc
      (local.get $0)
      (local.get $1)

**Comment by LeSeulArtichaut** [comment]
Just to clarify, this does not happen with LTO disabled?

**Comment by alexcrichton** [comment]
Ok I believe that this is fixed in https://github.com/rust-lang/rust/pull/82804. The `undefined` bit you see there is normal, that's what's happening with weak function references during LTO because that function isn't actually used.

**Comment by PiotrSikora** [comment]
> Just to clarify, this does not happen with LTO disabled?

Correct.

> Ok I believe that this is fixed in #82804. The `undefined` bit you see there is normal, that's what's happening with weak function references during LTO because that function isn't actually used.

Thanks! I verified that #82804 fixes the issue for me.

**Comment by apiraino** [comment]
Assigning `P-high` as discussed as part of the [Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2382758.20WASI.3A.20path_open.20regression.20in.20Rust.201.2E51) and removing `I-prioritize`.

@rustbot label -I-prioritize +P-high

**Comment by pnkfelix** [comment]
Discussed in [weekly rustc triage meeting](https://zulip-archive.rust-lang.org/238009tcompilermeetings/45595weekly2021040154818.html). We believe this was fixed by #82804. Closing as fixed.
